### DISCUSSION TITLE
API call not working

### DISCUSSION THREAD

**Discussion by ianww** [cables-gl/cables_docs#288] [Answers & Questions] [answered]
I'm new to both cables and APIs.

I'm trying to make an API call from this free RapidAPI site: [https://rapidapi.com/astronomyapi-astronomyapi-default/api/astronomy/](url)

Based on the code snippets on that site, in cables I've used ConcatMulti to concatenate the following strings, with the result being feed to AjaxRequest:

```
https://cors.cables.gl/ [I'VE TRIED IT BOTH WITH AND WITHOUT THIS CORS PROXY]
https://astronomy.p.rapidapi.com/api/v2/bodies/positions/
?X-RapidAPI-Key=730e563f72msh87c266b4ec899b8p15d469jsn6b82ee106270
&X-RapidAPI-Host=astronomy.p.rapidapi.com
&latitude=33.775867&longitude=-84.39733
&from_date=2017-12-20&to_date=2017-12-21
```

I'm getting a null return in AjaxRequest, and I can't work out why.  I'd really appreciate some advice!

**Comment by steam0r** [comment, marked as answer]
calling this url with ajax request gives me a 401 unauthorized error, so i guess something is wrong with your apikey or any other auth:

![Bildschirmfoto 2022-07-06 um 16 40 27](https://user-images.githubusercontent.com/151271/177577144-85d81196-24a8-42c0-8e61-ea5eac1e7663.png)

dev-tools really help to debug these networking issues

**Reply by ianww** [reply]
Thanks for that advice.  I've played around with it some more but not managed to get my authorisation right.  I suspect it is something to do with calling astronomyapi indirectly through rapidapi.  As I'm only trying to learn, I've changed over to use NASA APIs instead, which provide more understandable (for me) example calls, plus just requiring one API key for their whole site.  My calls to that site are working.  Yay!

**Reply by steam0r** [reply]
there is also https://www.postman.com/downloads/ for testing and debugging api, calls. sometimes thats more convenient than having to fiddle around with the browser dev-tool. glad to hear you got something working, though. congrats!

**Reply by ianww** [reply]
I hope that I don't have to pester for help node by node but I'm on a steep learning curve!  I'd appreciate some more assistance please...

OK, my AjaxRequest returned an object, and I'm now trying to extract an array of the cmeAnalyses key from it (the object has 3500 or so cmeAnalyses entries like this):

```
"cmeAnalyses": [
      {
        "time21_5": "2010-04-03T17:16Z",
        "latitude": 7,
        "longitude": 8,
        "halfAngle": 26,
        "speed": 620,
        "type": "C",
        "isMostAccurate": true,
        "note": "",
        "levelOfData": 0,
        "link": "https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CMEAnalysis/8677/-1",
        "enlilList": [
          {
            "modelCompletionTime": "2010-04-05T00:00Z",
            "au": 2,
            "estimatedShockArrivalTime": "2010-04-05T10:37Z",
            "estimatedDuration": 4.2,
            "rmin_re": 4.5,
            "kp_18": null,
            "kp_90": 3,
            "kp_135": 5,
            "kp_180": 6,
            "isEarthGB": false,
            "link": "https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/WSA-ENLIL/8678/-1",
            "impactList": [
              {
                "isGlancingBlow": false,
                "location": "STEREO A",
                "arrivalTime": "2010-04-05T19:00Z"
              },
              {
                "isGlancingBlow": true,
                "location": "Mars",
                "arrivalTime": "2010-04-07T11:00Z"
              }
            ],
            "cmeIDs": [
              "2010-04-03T09:54:00-CME-001"
            ]
          }
        ]
      },
      {
        "time21_5": "2010-04-03T14:11Z",
        "latitude": -25,
        "longitude": 3,
        "halfAngle": 32,
        "speed": 797,
        "type": "C",
        "isMostAccurate": false,
        "note": "Measured with swpc_cat.",
        "levelOfData": 0,
        "link": "https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CMEAnalysis/8676/-1",
        "enlilList": null
      }
    ],
```
I've connected an ObjectGetArray node to the AjaxRequest's `data` output, and I expected to see the ObjectGetArray fill with an array of 3500 or so entries of cmeAnalyses data but it stays null.

Am I doing something wrong or misunderstanding something?

**Reply by steam0r** [reply]
https://cables.gl/op/Ops.Json.ObjectGetArray_v2 should help you out, or you could use the structure view:

![Bildschirmfoto 2022-07-07 um 12 32 41](https://user-images.githubusercontent.com/151271/177753830-240bfad3-888a-4c5c-9b09-2cc64d8f6cff.png)

![Bildschirmfoto 2022-07-07 um 12 32 48](https://user-images.githubusercontent.com/151271/177753850-f37c6788-f5cb-4e09-ae6f-16f9c82a1f93.png)

there is also a series of tutorials on working with arrays in cables:

https://www.youtube.com/watch?v=FRFfvVgWFcs

**Reply by ianww** [reply]
Thanks, but that's what I'm using - https://cables.gl/op/Ops.Json.ObjectGetArray_v2 - and I'm getting a null return.  AjaxRequest is populated but ObjectGetArray is null. 

![ObjectGetArray](https://user-images.githubusercontent.com/36909839/177776290-036ce51a-d3c9-4f00-a6e8-60d72a6367f8.JPG)
![ObjectGetArray2](https://user-images.githubusercontent.com/36909839/177776320-80cd2bbc-b508-48de-b3b3-24112c3ec054.JPG)

**Reply by steam0r** [reply]
then thats maybe deeper nested and you need a few of these ops? cant really help any further without the datastructure or preferrably a public example patch, sorry

**Reply by ianww** [reply]
I've worked out that in my case ObjectGetArray was the wrong node to use to extract the data from the AjaxRequest.  Instead of ObjectGetArray I've successfully used ObjectValues to convert the object to an array, and then I used the structure gear icon to examine this array and help me select ArrayGetNumberByPath to extract a subarray of specific data.  So, now all good!

![image](https://user-images.githubusercontent.com/36909839/177967318-2a8d5b77-3535-47df-a61a-b2f49551b631.png)

![image](https://user-images.githubusercontent.com/36909839/177966632-d795fccd-8160-4f4a-9c6d-87aaddfa3c81.png)

Thanks again for your interest and advice.

**Reply by steam0r** [reply]
great, congratulations and also thanks for sharing your working result!
